karapace/schema_models.py

```diff
@@ -0,0 +1,159 @@
+
+    def to_json(self) -> Dict[str, Any]:
+        if self.schema_type is SchemaType.PROTOBUF:
+            raise InvalidSchema("Protobuf do not support to_json serialization")
+        return ujson.loads(self.schema_str)
+
+    def __str__(self) -> str:
+        if self.schema_type == SchemaType.PROTOBUF:
+            return self.schema_str
+        return json_encode(self.to_json(), compact=True)
+
+    def __repr__(self) -> str:
+        if self.schema_type == SchemaType.PROTOBUF:
+            return f"TypedSchema(type={self.schema_type}, schema={str(self)})"
+        return f"TypedSchema(type={self.schema_type}, schema={json_encode(self.to_json())})"
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, TypedSchema) and self.__str__() == other.__str__() and self.schema_type is other.schema_type
+
+
+class ValidatedTypedSchema(TypedSchema):
+    def __init__(self, schema_type: SchemaType, schema_str: str, schema: Union[Draft7Validator, AvroSchema, ProtobufSchema]):
+        super().__init__(schema_type=schema_type, schema_str=schema_str)
+        self.schema = schema
+
+    @staticmethod
+    def parse(schema_type: SchemaType, schema_str: str) -> "ValidatedTypedSchema":
+        if schema_type not in [SchemaType.AVRO, SchemaType.JSONSCHEMA, SchemaType.PROTOBUF]:
+            raise InvalidSchema(f"Unknown parser {schema_type} for {schema_str}")
+
+        parsed_schema: Union[Draft7Validator, AvroSchema, ProtobufSchema]
+        if schema_type is SchemaType.AVRO:
+            try:
+                parsed_schema = parse_avro_schema_definition(schema_str)
+            except (SchemaParseException, ValueError, TypeError) as e:
+                raise InvalidSchema from e
+
+        elif schema_type is SchemaType.JSONSCHEMA:
+            try:
+                parsed_schema = parse_jsonschema_definition(schema_str)
+                # TypeError - Raised when the user forgets to encode the schema as a string.
+            except (TypeError, ValueError, SchemaError, AssertionError) as e:
+                raise InvalidSchema from e
+
+        elif schema_type is SchemaType.PROTOBUF:
+            try:
+                parsed_schema = parse_protobuf_schema_definition(schema_str)
+            except (
+                TypeError,
+                SchemaError,
+                AssertionError,
+                ProtobufParserRuntimeException,
+                IllegalStateException,
+                IllegalArgumentException,
+                ProtobufError,
+                ProtobufException,
+                ProtobufSchemaParseException,
+            ) as e:
+                log.exception("Unexpected error: %s \n schema:[%s]", e, schema_str)
+                raise InvalidSchema from e
+        else:
+            raise InvalidSchema(f"Unknown parser {schema_type} for {schema_str}")
+
+        return ValidatedTypedSchema(schema_type=schema_type, schema_str=schema_str, schema=parsed_schema)
+
+    def __str__(self) -> str:
```

Comment:
This is overrides the base class `__str__` as there is a test for Protobuf schema field reordering.